In [9]:
import pandas as pd

df = pd.DataFrame({
    "study":[1,2,3,4,5,6,7,8],
    "sleep":[8,7,7,6,6,5,5,4],
    "homework":[2,4,6,8,10,12,14,16],
    "score":[52,58,65,70,76,82,90,95]
})

print(df)

   study  sleep  homework  score
0      1      8         2     52
1      2      7         4     58
2      3      7         6     65
3      4      6         8     70
4      5      6        10     76
5      6      5        12     82
6      7      5        14     90
7      8      4        16     95


In [10]:
# 2. 회귀모델 만들기

import statsmodels.api as sm

X = df[[
    "study",
    "sleep",
    "homework"
]]

y = df["score"]

X = sm.add_constant(X)
model = sm.OLS(y, X)
result = model.fit()

print(result.summary())



print(X.shape)
print(X.columns)

model = sm.OLS(y, sm.add_constant(X)).fit()

print(model.df_model)
print(model.df_resid)

                            OLS Regression Results                            
Dep. Variable:                  score   R-squared:                       0.999
Model:                            OLS   Adj. R-squared:                  0.998
Method:                 Least Squares   F-statistic:                     2103.
Date:                Tue, 16 Jun 2026   Prob (F-statistic):           4.86e-08
Time:                        19:50:54   Log-Likelihood:                -5.6012
No. Observations:                   8   AIC:                             17.20
Df Residuals:                       5   BIC:                             17.44
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         34.0500      7.481      4.552      0.0

# 선형회귀 결과 해석 실습 요약

## 회귀 결과

```text
Dep. Variable: score

R-squared: 0.999
Adj. R-squared: 0.998

coef      P>|t|

const      34.05   0.006
study       1.38   0.000
sleep       1.40   0.178
homework    2.76   0.000
```

---

# 1. 결정계수 (R²)

```text
R² = 0.999
```

의미

> 모델이 점수(score)를 99.9% 설명한다.

설명력이 매우 높음.

하지만

⚠️ R²가 높다고 반드시 좋은 모델은 아니다.

---

# 2. 수정 결정계수 (Adj R²)

```text
Adj R² = 0.998
```

의미

> 변수 개수를 고려한 설명력

R²와 거의 동일

→ 변수 수 과다 문제는 크지 않음

---

# 3. 전체 모델 유의성 (F-test)

```text
Prob(F-statistic)=4.86e-08
```

귀무가설

[
H_0:
모든 변수의 영향 = 0
]

결과

```text
p < 0.05
```

해석

> 모델 전체는 통계적으로 유의하다.

---

# 4. 변수별 해석

| 변수       | B(coef) | p-value | 해석            |
| -------- | ------: | ------: | ------------- |
| const    |   34.05 |   0.006 | 기본 점수         |
| study    |    1.38 |   0.000 | 공부 증가 → 점수 증가 |
| sleep    |    1.40 |   0.178 | 유의하지 않음       |
| homework |    2.76 |   0.000 | 과제 증가 → 점수 증가 |

---

## study

```text
coef = 1.38
p = 0.000
```

해석

공부시간 1 증가

↓

점수 약 1.38 증가

통계적으로 유의

---

## sleep

```text
coef = 1.40
p = 0.178
```

해석

수면시간 증가 시 점수 증가처럼 보이지만

```text
p > 0.05
```

통계적으로 유의하지 않음

→ 제거 고려

---

## homework

```text
coef = 2.76
p = 0.000
```

해석

과제시간 증가

↓

점수 증가

통계적으로 유의

---

# 5. 다중공선성 문제 발견

결과

```text
Cond. No. = 2.75e+16
```

경고

```text
strong multicollinearity problems
```

의미

독립변수끼리 매우 비슷함

예

```text
study      1 2 3 4
homework   2 4 6 8
```

거의 같은 정보

↓

모델이 구분 불가

---

문제점

* 계수가 불안정
* p-value 왜곡 가능
* 해석 어려움

---

해결 방법

```python
X=df[
[
"study",
"sleep"
]
]
```

중복 변수 제거 후 다시 학습

---

# 수업용 해석 순서

① R² 확인
→ 설명력 확인

↓

② F-test 확인
→ 모델 자체 유의성 확인

↓

③ p-value 확인
→ 의미 있는 변수 선택

↓

④ B(coef) 해석
→ 영향 방향 및 크기 확인

↓

⑤ 다중공선성 확인

---

# 최종 결론

* 설명력(R²)은 매우 높음
* study와 homework는 유의함
* sleep은 유의하지 않음
* study와 homework 사이 다중공선성이 매우 강함
* 변수 제거 후 재학습 필요
